fatoptimizer/inline.py

```diff
@@ -1,35 +1,128 @@
 import ast
 
-from .tools import OptimizerStep, pretty_dump
-#from .specialized import BuiltinGuard
+from .tools import OptimizerStep, pretty_dump, NodeTransformer, NodeVisitor
 
+class Checker(NodeVisitor):
+    '''Gather a list of problems that would prevent inlining a function.'''
+    def __init__(self):
+        self.problems = []
+
+    def visit_Call(self, node):
+        # Reject explicit attempts to use locals()
+        # TODO: detect uses via other names
+        if isinstance(node.func, ast.Name):
+            if node.func.id == 'locals':
+                self.problems.append('use of locals()')
```

Comment:
Hum, I started to implement something similar in namespace.py with ComplexAssignment, but your code is probably better.